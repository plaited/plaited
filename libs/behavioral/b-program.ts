import { strategies, streamEvents } from './constants.ts'
import { stateSnapshot } from './state-snapshot.ts'
import { createStream } from './create-stream.ts'
import { selectionStrategies } from './selection-strategies.ts'
import {
  CandidateBid,
  Feedback,
  ListenerMessage,
  LogCallback,
  ParameterIdiom,
  PendingBid,
  RulesFunc,
  RunningBid,
  Strategy,
  Trigger,
} from './types.ts'

const requestInParameter = (
  { event: requestEventName, detail: requestDetail = {} }: CandidateBid,
) => {
  return (
    {
      event: parameterEventName,
      cb: parameterAssertion,
    }: ParameterIdiom,
  ): boolean => (
    parameterAssertion
      ? parameterAssertion({
        detail: requestDetail,
        event: requestEventName,
      })
      : requestEventName === parameterEventName
  )
}

export const bProgram = ({
  /** event selection strategy {@link Strategy}*/
  strategy = strategies.priority,
  /** When set to true returns a stream with log of state snapshots, last selected event and trigger */
  dev = false,
}: {
  strategy?: Strategy | keyof Omit<typeof strategies, 'custom'>
  dev?: boolean
}) => {
  const eventSelectionStrategy: Strategy = typeof strategy === 'string'
    ? selectionStrategies[strategy]
    : strategy
  const pending = new Set<PendingBid>()
  const running = new Set<RunningBid>()
  const stream = createStream()

  function run() {
    running.size && step()
  }

  function step() {
    for (const bid of running) {
      const { bThread, priority, name } = bid
      const { value, done } = bThread.next()
      !done &&
        pending.add({
          name,
          priority,
          bThread,
          ...value,
        })
      running.delete(bid)
    }
    selectNextEvent()
  }
  // Select next event
  function selectNextEvent() {
    const bids = [...pending]
    const candidates = bids.reduce<CandidateBid[]>(
      (acc, { request, priority }) =>
        acc.concat(
          // Flatten bids' request arrays
          request && Array.isArray(request)
            ? request.map(
              (event) => ({ priority, ...event }), // create candidates for each request with current bids priority
            )
            : request
            ? [{ priority, ...request }]
            : [],
        ),
      [],
    )
    const blocked = bids.flatMap<ParameterIdiom>(({ block }) => block || [])

    const filteredBids: CandidateBid[] | never[] = candidates.filter(
      (request) => !blocked.some(requestInParameter(request)),
    )
    const selectedEvent = eventSelectionStrategy(filteredBids)
    if (selectedEvent) {
      dev && stream({
        type: streamEvents.snapshot,
        data: stateSnapshot({ bids, selectedEvent }),
      })
      const { priority: _p, cb: _a, ...detail } = selectedEvent
      stream({
        type: streamEvents.select,
        data: detail,
      })
      nextStep(selectedEvent)
    } else {
      stream({
        type: streamEvents.end,
        data: {
          strategy: typeof strategy === 'string' ? strategy : strategies.custom,
        },
      })
    }
  }
  // Queue up bids for next step of super step
  function nextStep(selectedEvent: CandidateBid) {
    for (const bid of pending) {
      const { request = [], waitFor = [], bThread } = bid
      const waitList = [
        ...(Array.isArray(request) ? request : [request]),
        ...(Array.isArray(waitFor) ? waitFor : [waitFor]),
      ]
      if (
        waitList.some(requestInParameter(selectedEvent)) && bThread
      ) {
        running.add(bid)
        pending.delete(bid)
      }
    }
    run()
  }
  const trigger: Trigger = ({
    event,
    detail = {},
  }) => {
    const bThread = function* () {
      yield {
        request: [{ event, detail }],
        waitFor: [{ event: '', cb: () => true }],
      }
    }
    running.add({
      name: event,
      priority: 0,
      bThread: bThread(),
    })
    if (dev) {
      const msg: ListenerMessage = {
        type: streamEvents.trigger,
        data: {
          event: event,
        },
      }
      detail && Object.assign(msg.data, { detail })
      stream(msg)
    }
    run()
  }

  const feedback: Feedback = (
    actions,
  ) => {
    stream.subscribe(
      ({ type, data }: ListenerMessage) => {
        if (type !== streamEvents.select) return
        const { event: key, detail = {} } = data
        type &&
          Object.hasOwn(actions, key) &&
          actions[key](detail)
      },
    )
  }

  const add = (logicStands: Record<string, RulesFunc>): void => {
    for (const name in logicStands) {
      running.add({
        name,
        priority: running.size + 1,
        bThread: logicStands[name](),
      })
    }
  }

  const log = (callback: LogCallback) => {
    stream.subscribe(
      ({ type, data }: ListenerMessage) => {
        if (type === streamEvents.trigger) {
          callback({ type, data })
        }
        if (type === streamEvents.snapshot) {
          callback({ type, data })
        }
        if (type === streamEvents.end) {
          callback({ type, data })
        }
      },
    )
  }
  return Object.freeze({
    /** add rule function to behavioral program */
    add,
    /** connect action function to behavioral program */
    feedback,
    /** trigger a run and event on behavioral program */
    trigger,
    /** reactive stream for logging selected events, state snapshots, and trigger events */
    log,
  })
}
