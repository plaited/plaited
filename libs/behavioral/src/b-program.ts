/* eslint-disable @typescript-eslint/no-use-before-define */
// import { stateSnapshot } from './state-snapshot.js'
import { ensureArray, isTypeOf } from '@plaited/utils'
import { priorityStrategy } from './selection-strategies.js'
import { publisher } from './publisher.js'
import {
  CandidateBid,
  DevCallback,
  Feedback,
  Parameter,
  PendingBid,
  RulesFunc,
  RunningBid,
  BPEvent,
  SnapshotMessage,
  Strategy,
  Trigger,
  BPEventTemplate,
} from './types.js'
import { loop, sync, thread } from './rules.js'
import { triggerWaitFor, log, isInParameter, isPendingRequest } from './utils.js'



/**
 * Creates a behavioral program that manages the execution of behavioral threads.
 * @param options An object containing optional parameters for the program.
 * @param options.strategy The event selection strategy to use. Defaults to `strategies.priority`.
 * @param options.dev A callback function that receives a stream of state snapshots, last selected event, and trigger.
 * @returns An object containing methods for managing the program and executing behavioral threads.
 */
export const bProgram = ({
  /** event selection strategy {@link Strategy}*/
  strategy = priorityStrategy,
  /** When set to true returns a stream with log of state snapshots, last selected event and trigger */
  dev: _dev,
}: {
  strategy?: Strategy
  dev?: DevCallback | true
} = {}) => {
  const dev = _dev === true ? log : _dev
  const pending = new Set<PendingBid>()
  const running = new Set<RunningBid>()
  const actionPublisher = publisher<BPEvent>()
  const snapshotPublisher = dev && publisher<SnapshotMessage>()
  function run() {
    running.size && step()
  }

  function step() {
    for (const bid of running) {
      const { generator, priority, thread, trigger } = bid
      const { value, done } = generator.next()
      !done &&
        pending.add({
          thread,
          priority,
          ...(trigger && { trigger }),
          generator,
          ...value,
        })
      running.delete(bid)
    }
    selectNextEvent()
  }
  // Select next event
  function selectNextEvent() {
    const blocked: Parameter[] = []
    const candidates: CandidateBid[] = []
    for (const bid of pending) {
      const { request, priority, block, thread } = bid
      block && blocked.push(...ensureArray(block))
      if (request) {
        Array.isArray(request) ?
          candidates.push(
            ...request.map(
              (event) => ({ priority, thread,  ...(isTypeOf<BPEventTemplate>(event, 'function') ? {template: event, ...event()} : event) }), // create candidates for each request with current bids priority
            ),
          )
        : candidates.push({ priority, thread, ...(isTypeOf<BPEventTemplate>(request, 'function') ? {template: request, ...request()} : request) })
      }
    }
    const filteredBids: CandidateBid[] = []
    const length = candidates.length
    for (let i = 0; i < length; i++) {
      const candidate = candidates[i]
       // Checking if candidate is in block Parameter
      if (!blocked.some(isInParameter(candidate))) {
        filteredBids.push(candidate)
      }
    }
    const selectedEvent = strategy(filteredBids)
    if (selectedEvent) {
      // snapshotPublisher && snapshotPublisher(stateSnapshot({ bids: [...pending], selectedEvent }))
      nextStep(selectedEvent)
    }
  }
  // Queue up bids for next step of super step
  function nextStep(selectedEvent: CandidateBid) {
    for (const bid of pending) {
      if (!bid.generator) continue
      if (
        // Checking if pending event is selectedEvent
        ensureArray(bid.request).some(isPendingRequest(selectedEvent)) ||
        // Checking if selectedEvent is in waitFor Parameter
        ensureArray(bid.waitFor).some(isInParameter(selectedEvent))
      ) {
        running.add(bid)
        pending.delete(bid)
      }
    }
    // To avoid infinite loop with calling trigger from feedback always publish select event
    // after checking if request(s) is waitList and before our next run
    actionPublisher({ type: selectedEvent.type, detail: selectedEvent.detail })
    run()
  }
  
  const trigger: Trigger = ({ type, detail }) => {
    const thread = function* () {
      yield {
        request: [{ type, detail }],
        waitFor: [triggerWaitFor],
      }
    }
    running.add({
      thread: type,
      priority: 0,
      trigger: true,
      generator: thread(),
    })
    run()
  }

  const feedback: Feedback = (actions) => {
    actionPublisher.subscribe((data: BPEvent) => {
      const { type, detail = {} } = data
      if (Object.hasOwn(actions, type)) {
        void actions[type](detail)
      }
    })
  }

  const addThreads = (threads: Record<string, RulesFunc>): void => {
    for (const thread in threads) {
      running.add({
        thread,
        priority: running.size + 1,
        generator: threads[thread](),
      })
    }
  }

  snapshotPublisher && snapshotPublisher.subscribe((data: SnapshotMessage) => dev(data))

  return Object.freeze({
    /** add thread function to behavioral program */
    addThreads,
    /** connect action function to behavioral program */
    feedback,
    /** trigger a run and event on behavioral program */
    trigger,
    /**
     * A behavioral thread that loops infinitely or until some callback condition is false
     * like a mode change open -> close. This function returns a threads
     */
    loop,
    /**
     * At synchronization points, each behavioral thread specifies three sets of events:
     * requested events: the threads proposes that these be considered for triggering,
     * and asks to be notified when any of them occurs; waitFor events: the threads does not request these, but
     * asks to be notified when any of them is triggered; and blocked events: the
     * threads currently forbids triggering
     * any of these events.
     */
    sync,
    /**
     * creates a behavioral thread from synchronization sets and/or other  behavioral threads
     */
    thread,
  })
}
