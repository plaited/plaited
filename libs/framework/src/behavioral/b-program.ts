import { isTypeOf } from '@plaited/utils'
import type {
  CandidateBid,
  SnapshotMessage,
  Feedback,
  BPListener,
  PendingBid,
  RunningBid,
  BPEvent,
  Trigger,
  BPEventTemplate,
  Rules,
  BProgram,
  Snapshot
} from './types.js'
import { triggerWaitFor, isListeningFor, isPendingRequest, createPublisher, ensureArray } from './private-utils.js'
import { snapshotFormatter } from './snapshot-formatter.js'
/**
 * Creates a behavioral program that manages the execution of behavioral threads.
 */
export const bProgram: BProgram = () => {
  const pending = new Map<string, PendingBid>()
  const running = new Map<string, RunningBid>()
  const actionPublisher = createPublisher<BPEvent>()
  let snapshotPublisher: {
    (value: SnapshotMessage): void;
    subscribe(listener: (msg: SnapshotMessage) => void | Promise<void>): () => void;
  } | undefined
  function run() {
    running.size && step()
  }

  function step() {
    for (const [thread, bid] of running) {
      const { generator, priority, trigger } = bid
      const { value, done } = generator.next()
      !done &&
        pending.set(thread, {
          priority,
          ...(trigger && { trigger }),
          generator,
          ...value,
        })
      running.delete(thread)
    }
    selectNextEvent()
  }
  // Select next event
  function selectNextEvent() {
    const blocked: BPListener[] = []
    const candidates: CandidateBid[] = []
    for (const [thread, { request, priority, block, trigger }] of pending) {
      block && blocked.push(...ensureArray(block))
      request &&
        candidates.push({
          priority,
          trigger,
          thread,
          ...(isTypeOf<BPEventTemplate>(request, 'function') ? { template: request, ...request() } : request),
        })
    }
    const filteredBids: CandidateBid[] = []
    const length = candidates.length
    for (let i = 0; i < length; i++) {
      const candidate = candidates[i]
      // Are we blocking the the candidate event
      if (!blocked.some(isListeningFor(candidate))) {
        filteredBids.push(candidate)
      }
    }
    /** @description Priority Queue BPEvent Selection Strategy */
    const selectedEvent = filteredBids.sort(
      ({ priority: priorityA }, { priority: priorityB }) => priorityA - priorityB,
    )[0]
    if (selectedEvent) {
      snapshotPublisher && snapshotPublisher(snapshotFormatter({ candidates, selectedEvent, pending }))
      nextStep(selectedEvent)
    }
  }
  // Queue up bids for next step of super step
  function nextStep(selectedEvent: CandidateBid) {
    for (const [thread, bid] of pending) {
      const { waitFor, request, generator } = bid
      if (!generator) continue

      if (
        // Is a pending a event the selectedEvent
        (request && isPendingRequest(selectedEvent, request)) ||
        // Are we waiting for selectedEvent
        ensureArray(waitFor).some(isListeningFor(selectedEvent))
      ) {
        running.set(thread, bid)
        pending.delete(thread)
      }
    }
    // To avoid infinite loop with calling trigger from feedback always publish select event
    // after checking if request(s) is waitList and before our next run
    actionPublisher({ type: selectedEvent.type, detail: selectedEvent.detail })
    run()
  }

  const trigger: Trigger = (request, triggerType) => {
    const thread = function* () {
      yield {
        request,
        waitFor: [triggerWaitFor],
      }
    }
    running.set(request.type, {
      priority: 0,
      trigger: triggerType ?? true,
      generator: thread(),
    })
    run()
  }

  const feedback: Feedback = (actions) => {
    actionPublisher.subscribe((data: BPEvent) => {
      const { type, detail = {} } = data
      if (Object.hasOwn(actions, type)) {
        void actions[type](detail)
      }
    })
  }

  const rules: Rules = {
    set: (threads) => {
      for (const thread in threads) {
        running.set(thread, {
          priority: running.size + 1,
          generator: threads[thread](),
        })
      }
    },
    has: (thread) => running.has(thread) || pending.has(thread),
    clear: () => {
      running.clear()
      pending.clear()
    },
    delete: (thread) => running.delete(thread) || pending.delete(thread),
  }

  const snapshot: Snapshot = listener => {
    if(snapshotPublisher === undefined) (snapshotPublisher = createPublisher<SnapshotMessage>())
    return snapshotPublisher.subscribe(listener)
  }

  return Object.freeze({
    /** add and delete rules functions of behavioral program */
    rules,
    /** connect action function to behavioral program */
    feedback,
    /** trigger a run and event on behavioral program */
    trigger,
    /** connect a listener that receives state snapshots for each step of a running behavioral program */
    snapshot,
  })
}
