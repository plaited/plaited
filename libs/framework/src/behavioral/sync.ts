import { isTypeOf } from '@plaited/utils'
import type { SynchronizationPoint, Synchronize } from './types.js'

/**
 * Creates a Rule Function from synchronization points.
 *
 * @param {...RulesFunction} args - An arbitrary number synchronization sets.
 * @returns {RulesFunction} A new behavioral thread that combines the provided synchronization sets.
 */
export const sync: Synchronize = (rules, repeat) => {
  return repeat ?
      function* () {
        while (isTypeOf<boolean>(repeat, 'boolean') ? repeat : repeat()) {
          const length = rules.length
          for (let i = 0; i < length; i++) {
            yield* rules[i]()
          }
        }
      }
    : function* () {
        const length = rules.length
        for (let i = 0; i < length; i++) {
          yield* rules[i]()
        }
      }
}

/**
 * Creates a synchronization point for a behavioral thread. At synchronization points,
 * each behavioral thread specifies three sets of parameters: request, waitFor, and block.
 *
 * @param {SynchronizationPoint} options - An object containing three properties: request, waitFor, and block. See [`SynchronizationPoint`](libs/behavioral/src/types.ts).
 * @param {BPEvent[] | BPEventTemplate} options.request - A behavioral program event or event template that the thread proposes. See [`BPEvent`](libs/behavioral/src/types.ts), [`BPEventTemplate`](libs/behavioral/src/types.ts).
 * @param {BPListener[] | BPListener} options.waitFor - An array of behavioral program listeners or a single listener that is waiting for a specific event or event who satisfies it's callback. See [`BPListener`](libs/behavioral/src/types.ts).
 * @param {BPListener[] | BPListener} options.blocked - An array of behavioral program listeners or a single listener that is blocking a specific event or event who satisfies it's callback. See [`BPListener`](libs/behavioral/src/types.ts).
 * @returns {RulesFunction} Synchronization set used for applying the execution of rules to the behavioral program. See [`RulesFunction`](libs/behavioral/src/types.ts).
 */
export const point: SynchronizationPoint = (syncPoint) =>
  function* () {
    yield syncPoint
  }
