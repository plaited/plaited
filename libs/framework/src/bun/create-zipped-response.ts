export const jsMimeTypes = ['text/javascript;charset=utf-8', 'application/json;charset=utf-8'] as const

export const isJsMimeType = (type: string): type is (typeof jsMimeTypes)[number] =>
  jsMimeTypes.includes(type as (typeof jsMimeTypes)[number])

export const createZippedResponse = (str: string, type: (typeof jsMimeTypes)[number]) => {
  const compressed = Bun.gzipSync(Buffer.from(str))
  return new Response(compressed, {
    headers: {
      'content-type': type,
      'content-encoding': 'gzip',
    },
  })
}
