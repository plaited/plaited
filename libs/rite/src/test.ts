/// <reference lib="dom" />
/// <reference lib="dom.iterable" />

import { TestResultError, TestResult, BrowserSessionResult } from '@web/test-runner-core/browser/session.js'
import { assert, Assertion, AssertionError } from './assert.js'

type TestCallback = (t: Assertion) => Promise<void> | void

interface Test {
  (name: string, cb: TestCallback): void
  skip: (name: string, cb: TestCallback) => void
}

type TestCase = () => Promise<boolean> | boolean

class TestRunner {
  #completedTests = 0
  #failedTests = 0
  #errors: TestResultError[] = []
  #cases: TestCase[] = []
  #timeout = 5_000
  #tests: TestResult[] = []
  updateTimeout(time: number) {
    this.#timeout = time
  }
  get timeout() {
    return this.#timeout
  }
  addCase(test: TestCase) {
    this.#cases.push(test)
  }
  addResult(result: TestResult) {
    this.#tests.push(result)
  }
  addError(obj: TestResultError) {
    this.#errors.push(obj)
  }
  async run(name: string) {
    for (const testCase of this.#cases) {
      try {
        const passed = await testCase()
        this.#completedTests++
        if (!passed) {
          this.#failedTests++
        }
      } catch (error) {
        console.error(error)
      }
    }

    if (this.#completedTests === this.#cases.length) {
      const results: BrowserSessionResult = {
        passed: !this.#failedTests,
        testResults: {
          name,
          suites: [],
          tests: this.#tests,
        },
        errors: this.#errors,
      }
      return results
    } else {
      throw new AssertionError(`Not all test completed`)
    }
  }
}

declare global {
  interface Window {
    __rite_test_runner?: TestRunner
  }
}

window.__rite_test_runner = new TestRunner()

const throwTimeoutError = async () => {
  const timeout = window.__rite_test_runner.timeout
  await assert.wait(timeout)
  throw new Error(`test takes longer than ${timeout}ms to complete`)
}

export const test: Test = (name, callback) => {
  window.__rite_test_runner.addCase(async () => {
    const startTime = performance.now()
    const result: TestResult = {
      name,
      passed: true,
      skipped: false,
    }
    try {
      await Promise.race([callback(assert), throwTimeoutError()])
      const msg = `✓ ${name}`
      // eslint-disable-next-line no-console
      console.log(msg)
    } catch (error) {
      const msg = `✗ ${name}`
      // eslint-disable-next-line no-console
      console.log(msg)
      result.passed = false
      if (error instanceof AssertionError) {
        const obj: Pick<TestResultError, 'actual' | 'expected' | 'message'> = JSON.parse(error.message)
        result.error = { name, stack: `${error.stack}`, ...obj }
        window.__rite_test_runner.addError({ name, ...obj })
      } else {
        throw new Error(error)
      }
    }
    result.duration = performance.now() - startTime
    window.__rite_test_runner.addResult(result)
    return result.passed
  })
}
test.skip = (name: string, _: TestCallback) => {
  window.__rite_test_runner.addCase(() => {
    const result = {
      name,
      passed: true,
      skipped: true,
    }
    window.__rite_test_runner.addResult(result)
    return true
  })
}
