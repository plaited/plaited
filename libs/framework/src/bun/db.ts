import { Database } from "bun:sqlite";

export const db = new Database(":memory:");
db.query(`create table templates;`);