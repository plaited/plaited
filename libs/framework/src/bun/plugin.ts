import type {  BunPlugin } from "bun";
import { db } from "./db.js";

export const plugin: BunPlugin = {
  name: "Plaited loader",
  setup(build) {
    build.onResolve({ filter: /\/route.tsx?$/ }, async ({ path }) => {
      // mark external all non-template files
    })
    build.onLoad({ filter: /\/route.tsx?$/ }, async ({ path }) => {
      // modify the contents of the file to be an export object.
      return {
        contents,
        loader: "tsx",
      };
    });
  },
}