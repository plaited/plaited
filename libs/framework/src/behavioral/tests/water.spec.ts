import { test, expect } from 'bun:test'
import { bProgram } from '../b-program.js'
import { point, sync } from '../sync.js'
import type { SnapshotMessage } from '../types.js'

test('Add hot water 3 times', () => {
  const actual: string[] = []
  const { bThreads, trigger, useFeedback } = bProgram()
  bThreads.set({
    addHot: sync([
      point({ request: { type: 'hot' } }),
      point({ request: { type: 'hot' } }),
      point({ request: { type: 'hot' } }),
    ]),
  })
  useFeedback({
    hot() {
      actual.push('hot')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual(['hot', 'hot', 'hot'])
})

test('Add hot/cold water 3 times', () => {
  const actual: string[] = []
  const { bThreads, trigger, useFeedback } = bProgram()
  bThreads.set({
    addHot: sync([
      point({ request: { type: 'hot' } }),
      point({ request: { type: 'hot' } }),
      point({ request: { type: 'hot' } }),
    ]),
    addCold: sync([
      point({ request: { type: 'cold' } }),
      point({ request: { type: 'cold' } }),
      point({ request: { type: 'cold' } }),
    ]),
  })
  useFeedback({
    hot() {
      actual.push('hot')
    },
    cold() {
      actual.push('cold')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual(['hot', 'hot', 'hot', 'cold', 'cold', 'cold'])
})

test('interleave', () => {
  const actual: string[] = []
  const { bThreads, trigger, useFeedback } = bProgram()
  bThreads.set({
    addHot: sync([
      point({ request: { type: 'hot' } }),
      point({ request: { type: 'hot' } }),
      point({ request: { type: 'hot' } }),
    ]),
    addCold: sync([
      point({ request: { type: 'cold' } }),
      point({ request: { type: 'cold' } }),
      point({ request: { type: 'cold' } }),
    ]),
    mixHotCold: sync([point({ waitFor: 'hot', block: 'cold' }), point({ waitFor: 'cold', block: 'hot' })], true),
  })
  useFeedback({
    hot() {
      actual.push('hot')
    },
    cold() {
      actual.push('cold')
    },
  })
  trigger({ type: 'start' })
  expect(actual).toEqual(['hot', 'cold', 'hot', 'cold', 'hot', 'cold'])
})

test('logging', () => {
  const snapshots: SnapshotMessage[] = []
  const { bThreads, trigger, useSnapshot } = bProgram()
  useSnapshot((snapshot: SnapshotMessage) => {
    snapshots.push(snapshot)
  })
  bThreads.set({
    addHot: sync([
      point({ request: { type: 'hot' } }),
      point({ request: { type: 'hot' } }),
      point({ request: { type: 'hot' } }),
    ]),
    addCold: sync([
      point({ request: { type: 'cold' } }),
      point({ request: { type: 'cold' } }),
      point({ request: { type: 'cold' } }),
    ]),
    mixHotCold: sync([point({ waitFor: 'hot', block: 'cold' }), point({ waitFor: 'cold', block: 'hot' })], true),
  })
  trigger({ type: 'start' })
  expect(snapshots).toMatchSnapshot()
})
