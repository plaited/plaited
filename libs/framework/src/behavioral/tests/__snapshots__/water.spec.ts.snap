// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`logging 1`] = `
[
  [
    {
      "blockedBy": undefined,
      "detail": undefined,
      "priority": 0,
      "selected": true,
      "thread": "start",
      "trigger": true,
      "type": "start",
    },
    {
      "blockedBy": undefined,
      "detail": undefined,
      "priority": 1,
      "selected": false,
      "thread": "addHot",
      "trigger": undefined,
      "type": "hot",
    },
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "priority": 2,
      "selected": false,
      "thread": "addCold",
      "trigger": undefined,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": undefined,
      "detail": undefined,
      "priority": 1,
      "selected": true,
      "thread": "addHot",
      "trigger": undefined,
      "type": "hot",
    },
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "priority": 2,
      "selected": false,
      "thread": "addCold",
      "trigger": undefined,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "priority": 1,
      "selected": false,
      "thread": "addHot",
      "trigger": undefined,
      "type": "hot",
    },
    {
      "blockedBy": undefined,
      "detail": undefined,
      "priority": 2,
      "selected": true,
      "thread": "addCold",
      "trigger": undefined,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": undefined,
      "detail": undefined,
      "priority": 1,
      "selected": true,
      "thread": "addHot",
      "trigger": undefined,
      "type": "hot",
    },
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "priority": 2,
      "selected": false,
      "thread": "addCold",
      "trigger": undefined,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "priority": 1,
      "selected": false,
      "thread": "addHot",
      "trigger": undefined,
      "type": "hot",
    },
    {
      "blockedBy": undefined,
      "detail": undefined,
      "priority": 2,
      "selected": true,
      "thread": "addCold",
      "trigger": undefined,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": undefined,
      "detail": undefined,
      "priority": 1,
      "selected": true,
      "thread": "addHot",
      "trigger": undefined,
      "type": "hot",
    },
    {
      "blockedBy": "mixHotCold",
      "detail": undefined,
      "priority": 2,
      "selected": false,
      "thread": "addCold",
      "trigger": undefined,
      "type": "cold",
    },
  ],
  [
    {
      "blockedBy": undefined,
      "detail": undefined,
      "priority": 2,
      "selected": true,
      "thread": "addCold",
      "trigger": undefined,
      "type": "cold",
    },
  ],
]
`;
